### PR TITLE
Add controller tests for pledges_controller #263

### PR DESCRIPTION
**Pull request by ashlynnpai**
I rule and have completed some work on Case Manager that's ready for review!

This pull request makes the following changes:
* The uncommented tests are passing
* The commented tests are failing but I put a note on each test as to why it's failing

It relates to the following issue #s: 
* Controller tests for pledges_controller #263

Sorry the whitespace and Gemfile.lock got messy. If it's too messy to read let me know.

### CODE REVIEW

File: test/controllers/pledges_controller_test.rb [hunk 1]
```diff
@@ -1,89 +1,99 @@
...
+  before do
+    @user = create :user
+    sign_in @user
+    @patient = create :patient
+    @pregnancy = create :pregnancy, patient: @patient
+  end
+
+  describe 'create method' do
+    before do
+      @pledge = attributes_for :pledge
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: 💯  great call on this!

File: test/controllers/pledges_controller_test.rb [hunk 2]
```diff
@@ -1,89 +1,99 @@
...
+
+#	NOTE: :pledge is getting passed in as a string
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: I think we can just remove this block that's commented out, since we're using `attributes_for :pledge` instead, which is severely cleaner. Thoughts?

File: test/controllers/pledges_controller_test.rb [hunk 3]
```diff
@@ -1,89 +1,99 @@
...
+      post :create, pregnancy_id: @pregnancy.id, pledge: @pledge
+    end
+
+    it 'should create and save a new pledge' do
+      assert_difference 'Pregnancy.find(@pregnancy).pledges.count', 1 do
+        post :create, pregnancy_id: @pregnancy.id, pledge: @pledge
+      end
+    end
+
+  # NOTE: it's redirecting so status code 302 (controller redirects to edit_pregnancy_path)
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: You are right on leaving these out for now.

File: test/controllers/pledges_controller_test.rb [hunk 4]
```diff
@@ -1,89 +1,99 @@
...
+#	NOTE: I created a blank template for this test	
+    it 'should render the correct template' do
+      assert_template 'pledges/update'
+    end
+
+    it 'should respond with success' do
+      assert_response :success
+    end
+
+  # NOTE: should work, but the call to reload is causing the factory attribute to reload,
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: INTERESTING. It sounds like there's something about the update method that's preventing it from actually saving -- we're running nearly identical test code for the notes controller and it seems to be working okay. 

Let me take a deeper look tomorrow.

File: test/factories/pledge.rb [hunk 5]
```diff
@@ -3,5 +3,6 @@
     pregnancy
     pledge_type 'Soft'
     amount 100
+    created_by { FactoryGirl.create(:user) }
```
Comment (colinxfleming) [inline test/factories/pledge.rb]: 💯

**Comment by colinxfleming** [comment]
@ashlynnpai you rule for pushing this over the finish line! I'm going to take a deeper look at some of the problems you ran into tomorrow, and we'll see if we can't get it hashed out. 

Thanks a ton!

File: test/controllers/pledges_controller_test.rb (hunk 2 above)
Comment (ashlynnpai) [inline test/controllers/pledges_controller_test.rb]: Yes, I wasn't sure if you wanted those specific values so I left them in. But if you don't need them.

File: test/controllers/pledges_controller_test.rb (hunk 4 above)
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: I think I found our sticking point! The problem child seems to be line 57, where we're building the update hash. I'll leave comments there.

File: test/controllers/pledges_controller_test.rb [hunk 6]
```diff
@@ -1,89 +1,99 @@
...
+
+    it 'should log the creating user' do
+      assert_equal Pregnancy.find(@pregnancy).pledges.last.created_by, @user
+    end
+  end		
+
+  describe 'update method' do		
+    before do
+      @pledge = create :pledge, pregnancy: @pregnancy, pledge_type: 'Original Pledge'
+      @pledge_edits = {pledge: {pledge_type: 'Edited Pledge'}}
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: Here's our sticking point! 

We have the `@pledge_edits` var as a nested hash here, right? Let's look at how we pass that in the `patch` command: 

`patch :update, pregnancy_id: @pregnancy, id: @pledge, pledge: @pledge_edits, format: :js`

So in effect, it feeds a payload that looks like this: ` { pledge: { pledge: { pledge_type: 'Edited Pledge' } } }`. Note the additional level of nesting -- in this, Rails is looking for an attribute called `pledge` on the pledge object, seeing that it doesn't exist, and just swallowing the rest of it. 

**In short, you should be able to make these tests pass by changing this line to `@pledge_edits = { pledge_type: 'Edited Pledge' }`.**

File: test/controllers/pledges_controller_test.rb [hunk 7]
```diff
@@ -1,89 +1,99 @@
...
+  #       assert_equal @pledge.history_tracks.count, 2
+  #       @changes = @pledge.history_tracks.last
+  #       assert_equal @changes.modified[:updated_by_id], @user.id
+  #       assert_equal @changes.modified[:pledge_type], 'Edited Pledge'
+  #     end
+
+  # NOTE: there is no model validatation (presence) for pledge type so having nil pledge type does not cause save to fail, 
+  # in contrast there is a model validation for note :full_text		
+  #     it 'should refuse to save pledge type to blank' do
+  #       [nil, ''].each do |bad_text|
```
Comment (colinxfleming) [inline test/controllers/pledges_controller_test.rb]: This test will probably fail as is, I don't think we're validating presence on pledge type in the model yet. You should feel free to make this change to get this test to pass as written. Let me know if you have any questions about that!

**Comment by colinxfleming** [comment]
Hey Ashlynn, I did that digging and I think I've found a solution. Give it a shot and let me know what you think?

**Comment by ashlynnpai** [comment]
Thanks for solving the pledge update bug! Incorporated your latest comments.

**Comment by colinxfleming** [comment]
:100: !

**Comment by colinxfleming** [comment]
This looks good to me! Thanks so much for sticking with this @ashlynnpai !
